src/features/textEditor/components/tipTapNodes/refStudioDocument/serialization/MarkdownSerializer.ts

```diff
@@ -0,0 +1,136 @@
+
+        if (!referenceId) {
+          return '@INVALID_REFERENCE';
+        }
+        const reference = this.referencesById[referenceId] as ReferenceItem | undefined;
+        if (reference) {
+          this.usedReferenceIds.add(referenceId);
+        }
+
+        return `@${reference?.citationKey ?? 'INVALID_REFERENCE'}`;
+      },
+    });
+  }
+
+  private inlineBlockToMarkdown(node: Node): string {
+    const nodeType = node.type.name;
+    if (nodeType !== 'paragraph' && nodeType !== 'heading') {
+      throw new Error('Unexpected node type. Expected an inline block, got ' + nodeType);
+    }
+
+    const htmlContent = getHTMLFromFragment(Fragment.from(node), this.editor.schema);
+    return this.turndownService.turndown(htmlContent);
+  }
+
+  private notionBlocksToMarkdown(nodes: Node[], indentLevel = 0, isList = false): string {
+    nodes.forEach((node) => {
+      const nodeType = node.type.name;
+      if (nodeType !== NotionBlockNode.name) {
+        throw new Error('Unexpected node type. Expected notionBlock, got ' + nodeType);
+      }
+    });
+
+    const lines: string[] = [];
+    let addEmptyLine = !isList;
+
+    nodes.forEach((node, nodeIndex) => {
+      const isCollapsible = node.attrs.type === 'collapsible' || node.attrs.type === 'bulletList';
+
+      if ((nodeIndex > 0 || indentLevel > 0) && (addEmptyLine || !isCollapsible)) {
+        lines.push('');
+      }
+
+      lines.push(' '.repeat(indentLevel * 4) + (isCollapsible ? '- ' : '') + this.inlineBlockToMarkdown(node.child(0)));
+
+      if (node.childCount > 1) {
+        const content: Node[] = [];
+        node.forEach((child, _offset, index) => {
+          if (index > 0) {
+            content.push(child);
+          }
+        });
+        lines.push(this.notionBlocksToMarkdown(content, isCollapsible ? indentLevel + 1 : indentLevel, isCollapsible));
+      }
+      addEmptyLine = !isCollapsible;
+    });
+    return lines.join('\n');
+  }
+
+  serialize(fileName: string): SerializedDocument {
+    this.usedReferenceIds = new Set<string>();
+
+    const { doc } = this.editor.state;
+
+    if (doc.childCount === 0) {
+      return { markdownContent: '' };
+    }
+    const notionBlocks: Node[] = [];
+    doc.forEach((child, _offset, index) => {
+      if (index > 0) {
+        notionBlocks.push(child);
+      }
+    });
+
+    const documentTitle = doc.child(0).textContent;
+
+    const markdownContent = [this.notionBlocksToMarkdown(notionBlocks)];
+
+    const metadata = ['---', `title: ${documentTitle}`, '---'];
+
+    if (this.usedReferenceIds.size === 0) {
+      return { markdownContent: [...metadata, '', ...markdownContent].join('\n') };
+    }
+
+    // Export references as .bib
+    const bibliography = serializeReferences(
+      [...this.usedReferenceIds.values()].map((referenceId) => this.referencesById[referenceId]).filter(Boolean),
+    );
+
+    metadata.splice(2, 0, `bibliography: ${fileName}.bib`);
```

Comment:
nit: I think that using splice is too dangerous here. I think it's ok to replace the metadata header for this scenario (with bib).